internal/proxy/client.go

```diff
@@ -0,0 +1,34 @@
+package proxy
+
+import (
+	"io"
+
+	abcicli "github.com/tendermint/tendermint/abci/client"
+	"github.com/tendermint/tendermint/abci/example/kvstore"
+	"github.com/tendermint/tendermint/abci/types"
+)
+
+// DefaultClientCreator returns a default ClientCreator, which will create a
+// local client if addr is one of: 'kvstore',
+// 'persistent_kvstore' or 'noop', otherwise - a remote client.
+//
+// The Closer is a noop except for persistent_kvstore applications,
+// which will clean up the store.
+func DefaultClientCreator(addr, transport, dbDir string) (abcicli.ClientCreator, io.Closer) {
```

Comment:
This is pre-existing, but is a deeply weird interface. Since it's constructing a factory anyway, why can't the closer be packaged into the factory and plumbed into the client? Of course, yes, it would have to be stashed until the creator actually gets invoked, but we're constructing a new instance for each call anyway.

We could do this separately, but we're already shaking things up in here, it might be a good time to improve the interface too.

Reply:
I'm less inclined to change this because it's in an internal package and is basically an implementation detail of the `node` package, but I think you're totally right.

Reply:
We don't necessarily have to, but having messy internal package APIs is only "better" than having messy external package APIs in that it's easier to fix them. 🙂  Actually now that I look at it, I think most of the "factory" business could probably be unexported, too. Maybe I'll do that as a separate PR all together once this merges.